### PR TITLE
[GOBBLIN-98] Orc records get dropped and duplicated

### PR DESCRIPTION
**Pull request by PraTrick**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title.
    - https://issues.apache.org/jira/browse/GOBBLIN-98

### Description
- [ ] Here are some details about my PR:
        The Hive's OrcSerde caches converted records - the OrcSerde has a single 
        org.apache.hadoop.hive.ql.io.orc.OrcSerde.OrcSerdeRow (basically an ArrayList) and every 
        time the org.apache.hadoop.hive.serde2.Serializer#serialize(Object, ObjectInspector) method 
        is called, the object is re-used.
        The problem is that org.apache.hadoop.hive.ql.io.orc.OrcSerde.OrcSerdeRow is package 
        protected and has no public constructor, so no copy can be made. This would be fine if 
        org.apache.hadoop.hive.ql.io.orc.OrcSerde.OrcSerdeRow is immediately written out. But all 
        Gobblin jobs have a buffer that the writer reads from. This buffering can cause race conditions 
        where records get dropped and duplicated.

### Tests
- [ ] My PR adds the following unit tests:
    - gobblin-core/src/test/java/gobblin/serde/HiveSerDeTest.java

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by abti** [comment]
@erwa @htran1 @aditya1105 Can either of you please review it, this PR has been sitting around for a while.

**Comment by abti** [comment]
ping @erwa @htran1 @aditya1105 ^^

File: gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ * @author Prateek Gupta
+ */
+
+    public class OrcSerDeWrapper extends OrcSerde {
+    private OrcSerde record = null;
+
+    @Override
+    public Writable serialize(Object realRow, ObjectInspector inspector) {
+        record = new OrcSerde();
+        Object realRowClone = ((ArrayList) realRow).clone();
```
Comment (erwa) [inline gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java]: How do you know `realRow` will be an `ArrayList`? I think you need to use Hive's `ObjectInspectorUtils.copyToStandardObject(realRow, inspector)` to copy the object and `ObjectInspectorUtils.getStandardObjectInspector(inspector)` to convert the OI accordingly as well.

File: gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+import org.apache.hadoop.hive.ql.io.orc.OrcSerde;
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspector;
+import org.apache.hadoop.io.Writable;
+
+import java.util.ArrayList;
+
+/**
+ * The Hive's {@link OrcSerde} caches converted records - the {@link OrcSerde} has a single
+ * {@link org.apache.hadoop.hive.ql.io.orc.OrcSerde.OrcSerdeRow} (basically an {@link ArrayList}) and every time the
```
Comment (erwa) [inline gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java]: The realRow inside the OrcSerdeRow might not be an ArrayList -- you should use the ObjectInspector to inspect the realRow; see comment below.

File: gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * @author Prateek Gupta
+ */
+
+    public class OrcSerDeWrapper extends OrcSerde {
+    private OrcSerde record = null;
+
+    @Override
+    public Writable serialize(Object realRow, ObjectInspector inspector) {
+        record = new OrcSerde();
```
Comment (erwa) [inline gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java]: No need to create a new `OrcSerde` here -- just call super.serialize() after cloning the object.

File: gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * The problem is that {@link org.apache.hadoop.hive.ql.io.orc.OrcSerde.OrcSerdeRow} is package protected and has no
+ * public constructor, so no copy can be made. This would be fine if {@link org.apache.hadoop.hive.ql.io.orc.OrcSerde.OrcSerdeRow}
+ * is immediately written out. But all Gobblin jobs have a buffer that the writer reads from. This buffering can cause
+ * race conditions where records get dropped and duplicated.
+ *
+ * @author Prateek Gupta
+ */
+
+    public class OrcSerDeWrapper extends OrcSerde {
```
Comment (erwa) [inline gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java]: unindent

File: gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java (hunk 1 above)
Comment (PraTrick) [reply gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java]: I've made the required changes in the latest commit, *except* the **ObjectInspector**. The OI works as expected.

File: gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java (hunk 2 above)
Comment (PraTrick) [reply gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java]: I've incorporated the changes as suggested in the latest commit.

File: gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java (hunk 3 above)
Comment (PraTrick) [reply gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java]: Changes accepted and done in the latest commit.

File: gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java (hunk 4 above)
Comment (PraTrick) [reply gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java]: Done.

**Comment by PraTrick** [comment]
@abti @erwa Thank you so much for the review. I've included changes as per your suggestion, and have also replied to your inline comments.

File: gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+ * race conditions where records get dropped and duplicated.
+ *
+ * @author Prateek Gupta
+ */
+
+public class OrcSerDeWrapper extends OrcSerde {
+
+  @Override
+  public Writable serialize(Object realRow, ObjectInspector inspector) {
+      Object realRowClone = ObjectInspectorUtils.copyToStandardObject(realRow, inspector);
```
Comment (erwa) [inline gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java]: Don't you need to import ObjectInspectorUtils, too?

File: gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+      return super.serialize(realRowClone, inspector);
```
Comment (erwa) [inline gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java]: The Javadoc for `ObjectInspectorUtils.copyToStandardObject()` says:
```
  /**
   * Returns a deep copy of the Object o that can be scanned by a
   * StandardObjectInspector returned by getStandardObjectInspector(oi).
   */
  public static Object copyToStandardObject(Object o, ObjectInspector oi) {
```

There's no guarantee that the `inspector` passed in is a `StandardObjectInspector`. For correctness, I think you need to use `ObjectInspectorUtils.getStandardObjectInspector(inspector)` here.

File: gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java (hunk 5 above)
Comment (PraTrick) [reply gobblin-core/src/main/java/org/apache/gobblin/converter/serde/OrcSerDeWrapper.java]: Oops... I think it got omitted while merging the changes. My apologies.

**Comment by PraTrick** [comment]
@abti @erwa All the changes have been completed. Please review the same. Thanks!

**Comment by erwa** [comment]
+1

**Pull request closed without merging**